### PR TITLE
Use HierarchicalSampler mixin with HMA

### LINKED ISSUE #1428: Make HMA use hierarchical sampling mixin
### Problem Description

_As an engineer, it would be helpful to have our synthesizers follow clear and well defined abstractions for the sake of modifying or enhancing functionality._

### Expected behavior

- Once #1411 is complete, make the `HMASynthesizer` inherit from that `BaseHierachicalSampler` and implement the required methods.
- Update the tests to match.

### Additional context

- All changes should be private. There shouldn't be a need to alter any user facing methods or classes.

### PR DESCRIPTION
**Pull request by frances-h**
Resolve #1428

### CODE REVIEW

File: sdv/sampling/hierarchical_sampler.py [hunk 1]
```diff
@@ -68,7 +68,7 @@ def _sample_rows(self, synthesizer, num_rows=None):
                 Sampled rows, shape (, num_rows)
         """
         num_rows = num_rows or synthesizer._num_rows
-        return synthesizer._sample_batch(num_rows, remove_extra_columns=False)
+        return synthesizer._sample_batch(int(num_rows), keep_extra_columns=True)
```
Comment (fealho) [inline sdv/sampling/hierarchical_sampler.py:71]: @amontanez24 @pvk-developer This line is an example where using an assert could be useful. `num_rows` should always be an integer, and if it is not then wherever it was set did so incorrectly, so the code there should be updated to cast it to an int.

So at this point we should `assert isinstance(num_rows, int)`, or change the docstring to say that `num_rows` can be a float as well. You shouldn't have to write code assuming other methods/inputs are invalid, ie casting to `int` in this case.

**Review by fealho — APPROVED** [review]
Looks good to me ⚡

File: sdv/multi_table/hma.py [hunk 2]
```diff
@@ -9,11 +9,12 @@
 from tqdm import tqdm
 
 from sdv.multi_table.base import BaseMultiTableSynthesizer
+from sdv.sampling import BaseHierarchicalSampler
 
 LOGGER = logging.getLogger(__name__)
 
 
-class HMASynthesizer(BaseMultiTableSynthesizer):
+class HMASynthesizer(BaseHierarchicalSampler, BaseMultiTableSynthesizer):
```
Comment (amontanez24) [inline sdv/multi_table/hma.py:17]: Does this order matter? If not could we flip it 😅

File: sdv/multi_table/hma.py [hunk 3]
```diff
@@ -484,89 +354,44 @@ def _get_likelihoods(self, table_rows, parent_rows, table_name, foreign_key):
...
-        else:
-            ratio = self._table_sizes[parent_name] / self._table_sizes[table_name]
-            num_parent_rows = max(int(round(len(table_rows) * ratio)), 1)
-            parent_model = self._table_synthesizers[parent_name]
-            parent_rows = self._sample_rows(parent_model, parent_name, num_parent_rows)
-
         primary_key = self.metadata.tables[parent_name].primary_key
-        parent_rows = parent_rows.set_index(primary_key)
-        num_rows = parent_rows[f'__{table_name}__{foreign_key}__num_rows'].fillna(0).clip(0)
+        parent_table = parent_table.set_index(primary_key)
```
Comment (amontanez24) [inline sdv/multi_table/hma.py:381]: @fealho referencing our discussion, this is an area where comments might be useful. Explaining why we set the index either here or in the docstring could be helpful

File: sdv/single_table/base.py [hunk 4]
```diff
@@ -604,10 +604,11 @@ def _sample_rows(self, num_rows, conditions=None, transformed_conditions=None,
                 * int:
                     Number of rows that are considered valid.
         """
-        if not self._random_state_set:
-            self._set_random_state(FIXED_RNG_SEED)
+        if (self._model and (
```
[the remark this answers is not in the document]
Comment (amontanez24) [reply sdv/single_table/base.py:607]: +1

File: sdv/single_table/base.py [hunk 5]
```diff
@@ -604,10 +604,11 @@ def _sample_rows(self, num_rows, conditions=None, transformed_conditions=None,
...
+                self._data_processor.get_sdtypes(primary_keys=False) or keep_extra_columns)):
+            if not self._random_state_set:
+                self._set_random_state(FIXED_RNG_SEED)
```
Comment (amontanez24) [inline sdv/single_table/base.py:610]: why can't we do this all the time anymore? Also can we add testing for this case?

**Review by amontanez24 — COMMENTED** [review]
Mostly minor comments! This is looking good :)

File: sdv/multi_table/hma.py (hunk 2 above)
Comment (frances-h) [reply sdv/multi_table/hma.py:17]: Order does matter, if we don't have `BaseHierarchicalSampler` first then it'll hit the `BaseMultiTableSynthesizer` methods _first_, so we'll end up hitting `NotImplemented` errors.

File: sdv/single_table/base.py (hunk 5 above)
Comment (frances-h) [reply sdv/single_table/base.py:610]: If we don't have a `_model` then this errors (the default implementation calls the `set_random_state` method on `self._model`.

**Review by amontanez24 — APPROVED** [review]
I think this looks good!

**Pull request merged**